beacon_node/http_api/src/publish_attestations.rs

```diff
@@ -0,0 +1,292 @@
+//! the first BN and therefore indicate that the attestations are
+//! "already seen". An attestation that has already been seen has
+//! been published on the network so there's no actual error from
+//! the perspective of the user.
+//!
+//! It's better to prevent slashable attestations from ever
+//! appearing on the network than trying to slash validators,
+//! especially those validators connected to the local API.
+//!
+//! There might be *some* value in determining that this attestation
+//! is invalid, but since a valid attestation already it exists it
+//! appears that this validator is capable of producing valid
+//! attestations and there's no immediate cause for concern.
+use crate::task_spawner::{Priority, TaskSpawner};
+use beacon_chain::{
+    validator_monitor::timestamp_now, AttestationError, BeaconChain, BeaconChainError,
+    BeaconChainTypes,
+};
+use beacon_processor::work_reprocessing_queue::{QueuedUnaggregate, ReprocessQueueMessage};
+use eth2::types::Failure;
+use lighthouse_network::PubsubMessage;
+use network::NetworkMessage;
+use slog::{debug, error, Logger};
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::{
+    mpsc::{Sender, UnboundedSender},
+    oneshot,
+};
+use types::Attestation;
+
+#[derive(Debug)]
+enum Error {
+    Validation(AttestationError),
+    Publication,
+    ForkChoice(BeaconChainError),
+    AggregationPool(AttestationError),
+    ReprocessDisabled,
+    ReprocessFull,
+    ReprocessTimeout,
+}
+
+enum PublishAttestationResult {
+    Success,
+    AlreadyKnown,
+    Reprocessing(oneshot::Receiver<Result<(), Error>>),
+    Failure(Error),
+}
+
+fn verify_and_publish_attestation<T: BeaconChainTypes>(
+    chain: &Arc<BeaconChain<T>>,
+    attestation: &Attestation<T::EthSpec>,
+    seen_timestamp: Duration,
+    network_tx: &UnboundedSender<NetworkMessage<T::EthSpec>>,
+) -> Result<(), Error> {
+    let attestation = chain
+        .verify_unaggregated_attestation_for_gossip(attestation, None)
+        .map_err(Error::Validation)?;
+
+    // Publish.
+    network_tx
+        .send(NetworkMessage::Publish {
+            messages: vec![PubsubMessage::Attestation(Box::new((
+                attestation.subnet_id(),
+                attestation.attestation().clone(),
+            )))],
+        })
+        .map_err(|_| Error::Publication)?;
+
+    // Notify the validator monitor.
+    chain
+        .validator_monitor
+        .read()
+        .register_api_unaggregated_attestation(
+            seen_timestamp,
+            attestation.indexed_attestation(),
+            &chain.slot_clock,
+        );
+
+    let fc_result = chain.apply_attestation_to_fork_choice(&attestation);
+    let naive_aggregation_result = chain.add_to_naive_aggregation_pool(&attestation);
+
+    if let Err(e) = fc_result {
+        Err(Error::ForkChoice(e))
+    } else if let Err(e) = naive_aggregation_result {
+        Err(Error::AggregationPool(e))
```

Comment:
We previously capture both errors - is it worth keeping both? i.e. returning a list of errors?

Reply:
I'll add some logs. I don't think the errors are particularly useful to send in the response, as the caller can't parse them and probably can't do anything. An error here means the attestation is really poor quality, or there is some serious bug.

Reply:
Logging added in 6043b0cca1674ee3b78099d3c82c863d76944a10